### ISSUE TITLE
 [BUG] `NaiveForecaster` performance issue when forecasting in-sample

### ISSUE THREAD

**Issue by kgeis** [alan-turing-institute/sktime#4460]
**Describe the bug**
I want to detrend before forecasting by normalizing my inputs to the latest value or to the mean within a window. It seems that I should be able to use `Detrender(forecaster=NaiveForecaster())` in a pipeline to achieve this. The performance makes it impossible. Using the default polynomial forecaster to detrend is much faster than using a constant forecaster.

**To Reproduce**
```python
from sktime.pipeline import make_pipeline
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.detrend import Detrender

import numpy as np
import pandas as pd
import xgboost as xgb

forecaster = make_pipeline(
    Detrender(),
    make_reduction(xgb.XGBRegressor())
)

y_train = pd.Series(np.random.rand(500), index=pd.date_range('1900-01-01', periods=500))
forecaster.fit(y_train) # takes ~0.4s

forecaster = make_pipeline(
    Detrender(forecaster=NaiveForecaster()),
    make_reduction(xgb.XGBRegressor())
)

forecaster.fit(y_train) # takes ~12s
```

**Expected behavior**
`Detrender(forecaster=NaiveForecaster())` should be as fast or faster than the default `Detrender(forecaster=PolynomialTrendForecaster(degree=1))`.

**Context**
MacBook Pro M1

**Versions**
<details>
System:
    python: 3.10.10 (main, Feb 16 2023, 02:46:59) [Clang 14.0.0 (clang-1400.0.29.202)]
executable: /Users/kgeis/temp/mystuff/sp_prob/bin/python
   machine: macOS-12.6.3-arm64-arm-64bit

Python dependencies:
          pip: 23.0.1
       sktime: 0.17.1
      sklearn: 1.2.2
        numpy: 1.23.5
        scipy: 1.10.1
       pandas: 1.5.3
   matplotlib: 3.7.1
       joblib: 1.2.0
  statsmodels: 0.13.5
        numba: 0.56.4
     pmdarima: 2.0.3
      tsfresh: 0.20.0
   tensorflow: None
tensorflow_probability: None
</details>

**Comment by fkiraly** [comment]
how strange. This does not look right.
Thanks for reporting!

**Comment by fkiraly** [comment]
same bug with only `sktime` code:

```python
from sktime.pipeline import make_pipeline
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.detrend import Detrender

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

forecaster = make_pipeline(
    Detrender(),
    make_reduction(LinearRegression())
)

y_train = pd.Series(np.random.rand(500), index=pd.date_range('1900-01-01', periods=500))
forecaster.fit(y_train) # takes ~0.1s
```

```python
forecaster = make_pipeline(
    Detrender(forecaster=NaiveForecaster()),
    make_reduction(LinearRegression())
)

forecaster.fit(y_train) # takes ~39s
```

**Comment by fkiraly** [comment]
verified on python 3.10, current `main`, windows

**Comment by fkiraly** [comment]
https://github.com/sktime/sktime/pull/4461 should fix `"last"`, but the other strategies are still slow.

Would be great to have some feedback from testing on `"last"` from you here, @kgeis.
If you can confirm it improved your use cases, I'd go ahead with the other strategies too.

**Comment by fkiraly** [comment]
re `"drift"`, first attempt here for a speed-up: https://github.com/sktime/sktime/pull/5072

review would be appreciated, @MCRE-BE!

Btw, I don't think using historical observed values is the right thing here, as those are not forecasts.

**Comment by fkiraly** [comment]
Speaking of which, @kgeis, have you ever gotten around to test the `"last"` strategy? Should be super fast now.

**Comment by MCRE-BE** [comment]
> Btw, I don't think using historical observed values is the right thing here, as those are not forecasts.
> 
> But the forecasts can be computed much more efficiently imo (attempt in the PR).

Just because I like to be annoying : why not? It does so for the other `NaiveForecaster` strategies and it says so in the documentation (or so I understand it).

_forecast by fitting a line between the first and last point of the window and extrapolating it into the future_

Why make it more complex than it needs to be?
